drv/cosmo-hf/src/hf.rs

```diff
@@ -0,0 +1,209 @@
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_page_program(
+            addr,
+            LeaseBufReader::<_, 32>::from(data.into_inner()),
+        )
+        .map_err(|()| RequestError::went_away())
+    }
+
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        offset: u32,
+        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+    ) -> Result<(), RequestError<HfError>> {
+        self.flash_read(
+            offset,
+            LeaseBufWriter::<_, 32>::from(dest.into_inner()),
+        )
+        .map_err(|_| RequestError::went_away())
+    }
+
+    fn sector_erase(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        protect: HfProtectMode,
+    ) -> Result<(), RequestError<HfError>> {
+        if addr as usize / SECTOR_SIZE_BYTES == 0
+            && !matches!(protect, HfProtectMode::AllowModificationsToSector0)
+        {
+            return Err(HfError::Sector0IsReserved.into());
+        }
+        self.flash_sector_erase(addr);
+        Ok(())
+    }
+
+    fn get_mux(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfMuxState, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_mux(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfMuxState,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn get_dev(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<HfDevSelect, RequestError<HfError>> {
+        todo!()
+    }
+
+    fn set_dev(
+        &mut self,
+        _: &RecvMessage,
+        _state: HfDevSelect,
+    ) -> Result<(), RequestError<HfError>> {
+        todo!()
+    }
+
+    fn hash(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        len: u32,
+    ) -> Result<[u8; SHA256_SZ], RequestError<HfError>> {
+        let hash_driver = drv_hash_api::Hash::from(HASH.get_task_id());
+        if hash_driver.init_sha256().is_err() {
+            return Err(HfError::HashError.into());
+        }
+        let begin = addr as usize;
+        // TODO: Begin may be an address beyond physical end of
+        // flash part and may wrap around.
+        let end = match begin.checked_add(len as usize) {
+            Some(end) => {
+                // Check end > maximum 4-byte address.
+                // TODO: End may be beyond physical end of flash part.
+                //       Use that limit rather than maximum 4-byte address.
+                if end > u32::MAX as usize {
+                    return Err(HfError::HashBadRange.into());
+                } else {
+                    end
+                }
+            }
+            None => {
+                return Err(HfError::HashBadRange.into());
+            }
+        };
+        // If we knew the flash part size, we'd check against those limits.
+        let mut buf = [0u8; PAGE_SIZE_BYTES];
```

Comment:
**question:** Is the idea that `being` is aligned at the start of a page and this can then read the flash part page-by-page (not counting the last, possibly not full page)? Is `start` guaranteed to be page-aligned?

Not sure if it would really make much of a difference, but you could split the range into full pages (with or without a not-full prefix, depending on if `start` is aligned or not) plus a not-necessarily-full suffix. That would mean you don't need to calculate `size` for the full pages.

Reply:
I think the comment on line 159 is slightly misplaced. What this code appears to be doing is just ensuring that reads are under 256 bytes in size by reusing a buffer; it doesn't appear to be sensitive to alignment, afaict.

Reply:
Yup, that comment isn't quite relevant here (we're limited by FPGA buffers, not the flash chip itself).  Deleted!